mountpoint-s3-fs/src/manifest/core.rs

```diff
@@ -24,213 +25,219 @@ pub enum ManifestError {
+        }
     }
 }
 
 impl TryFrom<DbEntry> for ManifestEntry {
     type Error = ManifestError;
 
     fn try_from(db_entry: DbEntry) -> Result<Self, Self::Error> {
-        match (db_entry.etag, db_entry.size) {
-            (None, None) => {
-                // no etag and no size means a directory entry
-                let parent_partial_key = db_entry.parent_partial_key.map(ValidKey::try_from).transpose()?;
-                if parent_partial_key.is_none() && db_entry.parent_id != ROOT_INODE_NO {
-                    Err(ManifestError::InvalidRow(db_entry.id))
-                } else {
-                    Ok(ManifestEntry::Directory {
-                        id: db_entry.id,
-                        parent_id: db_entry.parent_id,
-                        channel_id: db_entry.channel_id,
-                        parent_partial_key,
-                        name: db_entry.name,
-                    })
-                }
-            }
-            (Some(etag), Some(size)) => Ok(ManifestEntry::File {
-                // existing etag and size means a file entry
-                id: db_entry.id,
-                parent_id: db_entry.parent_id,
-                channel_id: db_entry.channel_id,
-                parent_partial_key: db_entry
-                    .parent_partial_key
-                    .map(ValidKey::try_from)
-                    .transpose()?
-                    .ok_or(ManifestError::InvalidRow(db_entry.id))?,
-                name: db_entry.name,
-                etag,
-                size,
-            }),
-            _ => Err(ManifestError::InvalidRow(db_entry.id)),
+        let parent_partial_key = db_entry.parent_partial_key.map(ValidKey::try_from).transpose()?;
+        if parent_partial_key.is_none() && db_entry.parent_id != ROOT_INODE_NO {
+            error!("only channel directories may have no parent_key, id: {}", db_entry.id);
+            return Err(ManifestError::InvalidRow(db_entry.id));
         }
+
+        let kind = match (db_entry.etag, db_entry.size) {
+            (None, None) => ManifestEntryKind::Directory,
+            (Some(etag), Some(size)) => ManifestEntryKind::File { etag, size },
+            _ => return Err(ManifestError::InvalidRow(db_entry.id)),
+        };
+
+        Ok(ManifestEntry {
+            id: db_entry.id,
+            parent_id: db_entry.parent_id,
+            channel_id: db_entry.channel_id,
+            parent_partial_key,
+            name: db_entry.name,
+            kind,
+            checksum: db_entry.checksum,
+        })
+    }
+}
+
+/// A helper to compute the checksum of the given set of fields.
+///
+/// Computes (partial_key, full_checksum, partial_checksum), with full_checksum including all of the fields
+/// and partial_checksum including only (partial_key, etag, size).
+///
+/// Note: this function will reconstruct the s3 key from parent_partial_key and name, this may mean unnecessary
```

Comment:
This is a bit vague. When do we need the copies and when are they wasted?

Reply:
Updated this comment to explain this and the difference between the "partial checksums" and the ones on the DB